go/vt/proto/vtgate/vstreamer/run.go

```diff
@@ -0,0 +1,464 @@
+			table.flushedRowCount += chunk.Len()
+			table.lastFlushedAt = time.Now()
+
+			v.tableFlushCount++
+			v.flushedRowCount += chunk.Len()
+		}
+
+		table.resetBatch()
+	}
+
+	// always store the latest vgtid, even if there are no customers to store
+	// err := storeLastVGtid(ctx, v.session, v.vgtidStateKeyspace, v.vgtidStateTable, v.latestVgtid)
+	// if err != nil {
+	// 	return err
+	// }
+
+	v.flushCount++
+	v.lastFlushedVgtid = v.latestVgtid
+	v.lastFlushedAt = time.Now()
+
+	return nil
+}
+
+func (table *TableConfig) resetBatch() {
+	if table.ReuseBatchSlice && table.currentBatch.IsValid() {
+		table.currentBatch = table.currentBatch.Slice(0, 0)
+	} else {
+		table.currentBatch = reflect.MakeSlice(reflect.SliceOf(reflect.PointerTo(table.underlyingType)), 0, table.MaxRowsPerFlush)
+	}
+}
+
+func (table *TableConfig) handleFieldEvent(ev *binlogdatapb.FieldEvent) error {
+	fieldMap := make(map[string]fieldMapping, len(ev.Fields))
+
+	for i := 0; i < table.underlyingType.NumField(); i++ {
+		structField := table.underlyingType.Field(i)
+		if !structField.IsExported() {
+			continue
+		}
+
+		// get the field name from the vstream, db, json tag, or the field name, in that order
+		mappedFieldName := structField.Tag.Get("vstream")
+		if mappedFieldName == "-" {
+			continue
+		}
+		if mappedFieldName == "" {
+			mappedFieldName = structField.Tag.Get("db")
+		}
+		if mappedFieldName == "" {
+			mappedFieldName = structField.Tag.Get("json")
+		}
+		if mappedFieldName == "" {
+			mappedFieldName = structField.Name
+		}
+
+		var found bool
+		for j, tableField := range ev.Fields {
+			if tableField.Name != mappedFieldName {
+				continue
+			}
+
+			found = true
+			fieldMap[mappedFieldName] = fieldMapping{
+				rowIndex:    j,
+				structIndex: structField.Index,
+				kind:        structField.Type.Kind(),
+				isPointer:   structField.Type.Kind() == reflect.Ptr,
+			}
+		}
+		if !found && table.ErrorOnUnknownFields {
+			return fmt.Errorf("vstreamer: field %s not found in provided data type", mappedFieldName)
+		}
+	}
+
+	// sanity check that we found at least one field
+	if len(fieldMap) == 0 {
+		return fmt.Errorf("vstreamer: no matching fields found for table %s", table.Table)
+	}
+
+	table.shards[ev.Shard] = shardConfig{
+		fieldMap: fieldMap,
+		fields:   ev.Fields,
+	}
+
+	return nil
+}
+
+func (table *TableConfig) handleRowEvent(ev *binlogdatapb.RowEvent) error {
+	shard, ok := table.shards[ev.Shard]
+	if !ok {
+		return fmt.Errorf("unexpected shard: %s", ev.Shard)
+	}
+
+	// grow the batch to the expected size
+	// table.currentBatch.Grow(len(ev.RowChanges))
+
+	for _, rc := range ev.RowChanges {
+		// for deletes, we'll leave the row null
+		//  TODO: this is wrong, it needs a PK
+		if rc.After == nil {
+			continue
+		}
+
+		row := sqltypes.MakeRowTrusted(shard.fields, rc.After)
+
+		// create a new struct for the row
+		v := reflect.New(table.underlyingType)
+		table.currentBatch = reflect.Append(table.currentBatch, v)
```

Comment:
While I really like the idea of returning a slice of their custom data type, it makes things like marking something as deleted tough to do. I'm considering returning a `[]Event` something like this, so there it still does all the work of assembling a row, but still lets you dive into the raw data if needed.

```go
type Event struct {
    Data any // this is their custom unmarshaled struct
    Deleted bool

    RawEvent *binlogdatapb.RowEvent
    RowChangeIndex // references the exact row change this was scanned from
}
```